File: preseason_edge_detection/colorspaces.py
# draws contours around certain HSV parameters, then draw a rectangle around the object
import cv2
import numpy as np


def nothing(x):
    pass

cap = cv2.VideoCapture(0)
cv2.namedWindow('frame1')
kernel = np.ones((5,5),np.uint8)

# create trackbars for color change
# lower
cv2.createTrackbar('HLo','frame1',19,179,nothing)
cv2.createTrackbar('SLo','frame1',57,255,nothing)
cv2.createTrackbar('VLo','frame1',123,255,nothing)
# upper
cv2.createTrackbar('HUp','frame1',52,179,nothing)
cv2.createTrackbar('SUp','frame1',197,255,nothing)
cv2.createTrackbar('VUp','frame1',255,255,nothing)

cv2.createTrackbar('areaTrackbar','frame1',10000,50000,nothing)

while(1):

	# Take each frame
	_, frame = cap.read()
	frame = frame[0:340, 60:610]

	# get current positions of four trackbars
	hLo = cv2.getTrackbarPos('HLo','frame1')
	sLo = cv2.getTrackbarPos('SLo','frame1')
	vLo = cv2.getTrackbarPos('VLo','frame1')
	hUp = cv2.getTrackbarPos('HUp','frame1')
	sUp = cv2.getTrackbarPos('SUp','frame1')
	vUp = cv2.getTrackbarPos('VUp','frame1')

	areaTrackbar = cv2.getTrackbarPos('areaTrackbar','frame1')



	hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

	# define range of interested (tuned to pink) color in HSV
	lower = np.array([hLo,sLo,vLo])
	upper = np.array([hUp,sUp,vUp])

	# Threshold the HSV image to get only blue colors
	mask = cv2.inRange(hsv, lower, upper)

	# Bitwise-AND mask and original image
	res = cv2.bitwise_and(frame,frame, mask= mask)

	opening = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
	closing = cv2.morphologyEx(opening, cv2.MORPH_CLOSE, kernel)
	laplacian = cv2.Laplacian(closing,cv2.CV_64F)
	sobelx = cv2.Sobel(closing,cv2.CV_64F,1,0,ksize=5)
	sobely = cv2.Sobel(closing,cv2.CV_64F,0,1,ksize=5)

	cv2.imshow("closing", closing)


	# contours!
	contours = cv2.findContours(closing,cv2.cv.CV_RETR_TREE,cv2.cv.CV_CHAIN_APPROX_SIMPLE)[0]

	approx_contours = []

	for c in contours:
		moments = cv2.moments(c)
		center = (int(moments['m10']/moments['m00']), (int (moments['m01']/moments['m00'])))
		area = cv2.contourArea(c)
		if area < areaTrackbar: continue
		perim = cv2.arcLength(c, True)
		approx = cv2.approxPolyDP(c, 0.00005*perim, True)

		if len(approx) > 20:
			approx_contours.append(approx)


		for x in range(0, len(approx_contours)):
			minX = 10000
			minY = 10000
			maxX = 0
			maxY = 0
			cx = (minX + maxX)//2
			cy = (minY + maxY)//2

			for y in range(0, len(approx_contours[x])):
				if approx_contours[x][y][0][0] < minX:
					minX = approx_contours[x][y][0][0]
				if approx_contours[x][y][0][1] < minY:
					minY = approx_contours[x][y][0][1]
				if approx_contours[x][y][0][0] > maxX:
					maxX = approx_contours[x][y][0][0]
				if approx_contours[x][y][0][1] > maxY:
					maxY = approx_contours[x][y][0][1]
			if approx_contours:
				cv2.drawContours(frame, approx_contours, -1, (255, 0, 0),3)
				cv2.rectangle(frame,(minX, minY),(maxX, maxY),(255,255,0),3)

			percentFull = ((340 - minY) / 340.0) * 100
			print(percentFull)


	cv2.imshow('frame',frame)

	k = cv2.waitKey(5) & 0xFF
	if k == 27:
		break

cv2.destroyAllWindows()

File: fuel_detection_counter/colorspaces.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <math.h>

#include "C920Camera.h"

using namespace cv;
using namespace std;
class FuelDetector {

		Scalar min;
		Scalar max;
		int minArea;
		float h;
		vector<vector<Point> > out;
		int area;
		Mat hsv;
		Mat mask;
		Mat opening;
		Mat closing;
		Mat res;
		Mat kernel;
		vector<vector<Point> > contours;

	public: FuelDetector(float he) {
		max = Scalar(0,255,255);
		minArea=100;
		h=he;
		kernel = Mat(5,5, CV_8U);
	}

	void changeMin(int hLo,int hUp,int sLo,int vLo) {
		min = Scalar(hLo,sLo,vLo);
		max.val[0] = hUp;
	}

	vector<vector<Point> > getFuel(Mat frame)
	{
		out.clear();	//reset the output list
		cvtColor(frame, hsv, CV_BGR2HSV);	//create an hsv image for filtering
		inRange(hsv, min, max, mask);	//create the mask
		bitwise_and(frame,frame, res, mask=mask);
		morphologyEx(mask, opening, MORPH_OPEN, kernel);	//expand the mask
		morphologyEx(opening, closing, MORPH_CLOSE, kernel);	//contract the mask
		//contours
		findContours(closing,contours,RETR_TREE,CHAIN_APPROX_SIMPLE);
		if (contours.size()>0) {	//check is there are any contours
			for (int x = 0; x < contours.size(); x++) {	//for each contours do stuff
				Rect rec = boundingRect(contours[x]);
				if (contourArea(contours[x]) > minArea && rec.width*rec.height < (frame.rows-40)*(frame.cols-40) ) {	//check to see the size of the contours
					out.push_back(contours[x]);	//add contours to output list
				}
			}
		}
		return out;
	}

	float angleToDist(vector<Point> c) {
		if (c.size() == 0) throw invalid_argument("received empty contour");
		float a=.0010496;	//angle of each pixel .0010496 for c920 at 720p
		float b=1.54; //angle from top of view of camera to straight down in radians
		int sumOfY=0;
		for (int i = 0; i < c.size(); i++) {
			sumOfY+=c[i].y;
		}
		return h*tan(b-a*(sumOfY/c.size()));
	}



	float expectedSize(float dist) {
		float sizeConstant=6430;		//try 4669
		return sizeConstant/sqrt(pow(h,2)+pow(dist,2));

	}

	float fuelCount(vector<Point> c) {
		if (c.size() == 0) throw invalid_argument("received empty contour");
		double dist = angleToDist(c);
		double eSize = expectedSize(dist);
		return contourArea(c)/eSize;
	}

	float angleToCluster(vector<Point> c) {
		float x =.0009602;
		int sumOfX=0;
		for(int i = 0; i < c.size(); i++) {
			sumOfX+=c[i].x;
		}
		return (-x*640+x*sumOfX/c.size());
	}

};



int main(int, char**)
{
	v4l2::C920Camera camera(1);
	camera.SetBrightness(60);
	camera.SetWhiteBalanceTemperature(30);
	camera.SetGain(45);
	camera.SetSaturation(190);
	camera.SetContrast(130);
	camera.ChangeCaptureSize(v4l2::CAPTURE_SIZE_1280x720);
	//VideoCapture cap(1); // open the default camera
    if(!camera.IsOpen())  // check if we succeeded
        return -1;


    namedWindow("frame",1);

	// create trackbars for color change
	// lower
	int hLo = 29;
	int sLo = 65;
	int vLo = 80;
	int hUp = 38;
	//int sUp = 255;
	//int vUp = 255;
	//int areaTrackbar = 10000;

	createTrackbar("HLo","frame",&hLo,179);
	createTrackbar("SLo","frame",&sLo,255);
	createTrackbar("VLo","frame",&vLo,255);
	createTrackbar("HUp","frame",&hUp,179);
	/*
	// upper
	createTrackbar("SUp","frame",&sUp,255);
	createTrackbar("VUp","frame",&vUp,255);

	createTrackbar("areaTrackbar","frame",&areaTrackbar,50000);
*/
	double total;
	vector<vector<Point> > fuel;
	Mat frame;
	FuelDetector b = FuelDetector(.70);
	b.changeMin(hLo,hUp,sLo,vLo);
	while (true) {
		camera.GrabFrame();
		camera.RetrieveMat(frame);
		//cap >> frame; // get a new frame from camera

		// get current positions of four trackbars
		/*
		sUp = getTrackbarPos("SUp","frame");
		vUp = getTrackbarPos("VUp","frame");

		areaTrackbar=getTrackbarPos("areaTrackbar","frame");

		Scalar	min(hLo, sLo, vLo);
	    Scalar	max(hUp, sUp, vUp);
*/
		fuel = b.getFuel(frame);
		total=0;
		drawContours(frame,fuel,-1,Scalar(255,0,0),2);
		for (int i = 0; i < fuel.size(); i++) {
			total += b.fuelCount(fuel[i]);
			cout << "Dist " << i << ": " << b.angleToDist(fuel[i]) << " ";
		}
		cout <<"Fuel:" << total  << endl;
		imshow("frame", frame);
        if(waitKey(30) >= 0) break;
    }
    // the camera will be deinitialized automatically in VideoCapture destructor
    return 0;
}

